Unbeatable computer

lpenzey opened the pull request:
User can now select from three game modes:

1.  Human vs. human
2. Human vs. easy computer
3. Human vs unbeatable computer (minimax)

Also included is the ability for the user to select an alphanumeric/emoji character for their player symbol and the computer's symbol.

lpenzey committed:
User can choose alphanumeric player token

The user can now choose any alphanumerical character for their player
token. Status calculates a winning combination if all the elements in
the combination are the same and doesn't care about what the actual
character is.

This change necessitated adding two new fields to the state struct:
opponent and current player because there are no default player tokens,
and thus switching between players requires distinct struct fields. This
also adds the ability for the game to know who's turn it is currently
and block attempts to place a move by a user that isn't the current
player.

The only character not allowed is if the opponent's token is identical
to the player's token.

If it is the same the user will be prompted again to enter a unique
token.

lpenzey committed:
Add unbeatable computer player

The unbeatable computer implements the minimax algorithm to ensure that
the worst outcome for the computer player is a tie. It also handles
cases where the player and opponent symbols are numbers that are also
positions on the board, since the symbols are represented as strings and
spots on the board are integers.

lpenzey committed:
Confirm result of calling win? on tie game

lpenzey committed:
Make move prompt dynamic with player's token

lpenzey committed:
Tweak set move function

lpenzey committed:
Add choice of hard or easy computer opponent

ssunday commented on lib/TTT/game.ex:
```diff
@@ -1,36 +1,48 @@
+    play(state, deps, mode)
   end
 
   def select_mode(mode) when mode == "1", do: :human_v_human
```
Constants for the mode numbers?

lpenzey replied:
smart 👍

ssunday commented on lib/TTT/game.ex:
```diff
@@ -44,12 +56,29 @@ defmodule Game do
+  def hard_computer_move(state) do
+    move =
+      try do
+        Minimax.minimax(state).position
```
It feels like this could be merged—the rescue and the case since it comes right after each other.

ssunday commented on lib/TTT/minimax.ex:
```diff
@@ -0,0 +1,49 @@
+
+  def score_move(state, winner) do
+    cond do
+      state.player == winner ->
```
(Not sure if this would work, but could you use a case or check based on the value of winner?)

lpenzey replied:
you could but you'd have to create the variables first to match against winner because you can't invoke remote function in matches (like what I did in `optimal_move`, but I'm not sure if there's a preference or if one is more idiomatic)

ssunday replied:
Could you deconstruct state and use variable pattern matching in the function?

I'm not 100% on what's better/easily possible, but this works

lpenzey committed:
Make game modes module attributes

rmulhol commented on lib/TTT/game.ex:
```diff
@@ -1,36 +1,51 @@
 defmodule Game do
```
Heads up that when I went to run this, I got the follow error:
```
mix run -e 'TicTacToe.main()'
Unchecked dependencies for environment dev:
* mix_test_watch (Hex package)
  the dependency is not available, run "mix deps.get"
** (Mix) Can't continue due to errors on dependencies
```
Pretty self-explanatory, but might be worth adding a `deps` section to the README

rmulhol commented on lib/TTT/minimax.ex:
```diff
@@ -0,0 +1,49 @@
+    end
+  end
+
+  def optimal_move(moves, state) do
```
on the subject of "optimal", seems like the computer player is opting to block a loss instead of taking an available win:
![Screen Shot 2019-05-10 at 12 31 15 PM](https://user-images.githubusercontent.com/7875345/57545375-b9ce5b80-731f-11e9-97c1-339644002e9e.png)
How could we tune the algorithm so that the computer would prefer winning over simply not losing?

rmulhol commented on lib/TTT/player.ex:
```diff
@@ -18,17 +18,20 @@ defmodule Player do
+  def choose_token(deps, player_token) do
+    token =
+      deps.io.retrieve(Messages.get_message(:choose_opponent_token))
+      |> Validation.clean()
```
I imagine we might want to add some more constraints on what's allowed as a valid token :)

![Screen Shot 2019-05-10 at 12 51 37 PM](https://user-images.githubusercontent.com/7875345/57546453-614c8d80-7322-11e9-8960-19f26154fb63.png)

rmulhol reviewed (commented):
nice work! some UI comments follow...

lpenzey committed:
Merge try/rescue and case statement

lpenzey committed:
Fix missing user prompt message

lpenzey committed:
Refactor out unnecessary if statement

lpenzey committed:
Require player token to be one character long

Validator checks to see if the token the user enters is one byte long,
and asks for another token if the given token is less or more than one
byte long. Checking for length of the player token in this way removes
capability to have emojis as player tokens since they're 4 bytes long. A
custom guard to check for the length of a string, not the number of
bytes, would allow for emojis to be entered.

lpenzey committed:
Replace deps  with io_wrapper

Deps previously was a map with multiple dependencies, but since most of
those "dependencies" were internally owned modules, the only dependency
left is on Elixir's built in IO module. Thus deps has been replaced with
a single module, the io wrapper module TTT.IO.

Now when TicTacToe.main() calls the options builder it returns the IO
wrapper (TTT.IO), and the io wrapper accesses the config file to set
Elixir's IO module.

lpenzey committed:
Update readme with build instructions

ssunday commented on lib/TTT/validation.ex:
```diff
@@ -35,10 +35,47 @@ defmodule Validation do
+  end
+
+  def validate_token(opponent_token, player_token, _io)
+      when opponent_token != player_token and byte_size(opponent_token) == 1 do
```
spacing seems weird?

lpenzey replied:
yeah it is weird, the auto-formatter thinks the guard clause is too long so it puts it on a new line

ssunday commented on lib/tic_tac_toe.ex:
```diff
@@ -1,6 +1,6 @@
-    deps = OptionsBuilder.init()
-    Game.start(deps)
+    io = OptionsBuilder.init()
+    Game.start(io)
```
OptionsBuilder.init() |> Game.start?

lpenzey committed:
Make minimax prefer win over guaranteeing tie

Minimax will now secure a win for itself when presented with the option
to block the opposing player or score a winning combination. This was
done by adjusting the payout function that scores wins and losses for
the min and max players.

lpenzey committed:
Rename Player.analyze for clarity

Player.analyze has become Player.check_move since it's a better
description of what the function is focused on accomplishing.

lpenzey committed:
Simplify play function

Game.play has been refactored to only have one function definition.
Different modes are enabled by the opponent() abstraction, which selects
the correct opponent based on the game mode atom.

lpenzey committed:
Pipe result of OptionsBuilder to Game.start

rmulhol commented on lib/TTT/minimax.ex:
```diff
@@ -36,13 +40,21 @@ defmodule Minimax do
       ^min ->
-        Enum.reduce(moves, %{player: nil, position: nil, score: 10}, fn move, acc ->
-          if move.score <= acc.score, do: move, else: acc
+        Enum.reduce(moves, %{position: nil, score: 10}, fn move, acc ->
```
Interesting to me that this namespace is using `20`/`-20` _and_ `10`/`-10`. Are these [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming))? I'm a little bit confused about how much I could change these numbers before tests would start failing.
